src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js

```diff
@@ -1,171 +1,560 @@
+/**
+ * Helper function to load scripts.
+ *
+ * It runs only once, so it won't load the same script on this or any other visualization show/hide.
+ */
 function loadScript(url) {
-    var script = document.createElement("script");
+    let script = document.createElement("script");
     script.src = url;
 
     document.head.appendChild(script);
 }
 
-loadScript('https://d3js.org/d3.v4.min.js');
+/**
+ * Helper function to load styles.
+ *
+ * It runs only once, so it won't load the same style on this or any other visualization show/hide.
+ */
+function loadStyle(url) {
+    let style   = document.createElement("link");
+    style.href  = url;
+    style.rel   = "stylesheet";
+    style.media = "screen";
+    style.type  = "text/css";
+
+    document.head.appendChild(style);
+}
+
+const label_style        = "font-family: DejaVuSansMonoBook; font-size: 10px;";
+const x_axis_label_width = 30;
+const label_padding_x    = 7;
+const label_padding_y    = 2;
+const animation_duration = 1000;
 
 /**
  * A d3.js ScatterPlot visualization.
  *
- * source (CSV file): | x-axis | y-axis | domain |
+ * To zoom use scrollwheel
+ * To select click and swipe with LMB
+ * To deselect click outside of selection with LMB
+ * To pan click and swipe with RMB
+ * To zoom out click "Fit all" or use key combination "ctrl/cmd+a"
+ * To zoom into selection click appropriate button or use key combination "ctrl/cmd+z"
+ *
  * Data format (json):
  * {
- *  source : String,
- *  colors : [{
- *      domain : String,
- *      color  : String
- *  }]
+ *  "axis":{
+ *     "x":{"label":"x-axis label""scale":"linear"},
+ *     "y":{"label":"y-axis label","scale":"logarithmic"},
+ *  },
+ *  "focus":{"x":1.7,"y":2.1,"zoom":3.0},
+ *  "points":{"labels":"visible" | "invisible"},
+ *  "data":[
+ *     {"x":0.1,"y":0.7,"label":"foo","color":"FF0000","shape":"circle","size":0.2},
+ *     ...
+ *     {"x":0.4,"y":0.2,"label":"baz","color":"0000FF","shape":"square","size":0.3}
+ *  ]
  * }
  */
 class ScatterPlot extends Visualization {
     static inputType = "Any"
-    static label = "Scatter Plot (JS)"
+    static label     = "Scatter Plot (JS)"
 
+    /**
+     * Presents a scatterplot visualization after receiving `data`.
+     */
     onDataReceived(data) {
+        loadScript('https://d3js.org/d3.v4.min.js');
+        loadStyle('https://fontlibrary.org/face/dejavu-sans-mono')
+
         while (this.dom.firstChild) {
             this.dom.removeChild(this.dom.lastChild);
         }
 
-        let width     = this.dom.getAttributeNS(null, "width");
-        let height    = this.dom.getAttributeNS(null, "height");
-        const divElem = this.createDivElem(width, height);
+        let width        = this.dom.getAttributeNS(null,"width");
+        let height       = this.dom.getAttributeNS(null,"height");
+        const btnPadding = 25;
```

Comment:
what kind of apdding this is? The var does not tell wherther this is top, bottom padding and from what element